src/proposer/proposer_rpc.cpp

```diff
@@ -167,6 +170,125 @@ class ProposerRPCImpl : public ProposerRPC {
     obj.pushKV("stakeable_coins", arr);
     return obj;
   }
+
+  UniValue ProposeBlocks(CWallet &wallet,
+                         const boost::optional<CScript> &coinbase_script,
+                         const int num_generate) const {
+
+    proposer::Proposer *proposer = GetComponent<proposer::Proposer>();
+
+    UniValue block_hashes(UniValue::VARR);
+
+    // To pick up to date coins for staking we need to make sure that the wallet is synced to the current chain.
+    wallet.BlockUntilSyncedToCurrentChain();
+
+    esperanza::WalletExtension &wallet_ext = wallet.GetWalletExtension();
+
+    for (int i = 0; i < num_generate; ++i) {
+
+      std::shared_ptr<const CBlock> block;
+      {
+        LOCK2(cs_main, wallet_ext.GetLock());
+        const staking::CoinSet stakeable_coins = wallet_ext.GetStakeableCoins();
+
+        if (stakeable_coins.empty()) {
+          throw JSONRPCError(RPC_INTERNAL_ERROR,
+                             "Not proposing, not enough balance.");
+        }
+
+        // We don't want to combine coins when we use the rpc, so we don't need
+        // to pass all of them.
+        const staking::CoinSet first_coin = {*stakeable_coins.begin()};
+
+        block = proposer->GenerateBlock(wallet,
+                                        first_coin,
+                                        coinbase_script);
+        if (!block) {
+          throw JSONRPCError(RPC_INTERNAL_ERROR, "Failed to generate a block.");
+        }
+      }
+
+      if (!ProcessNewBlock(Params(), block, /* fForceProcessing= */ true, nullptr)) {
+        throw JSONRPCError(RPC_INTERNAL_ERROR, "ProcessNewBlock, block not accepted");
+      }
+
+      block_hashes.push_back(block->GetHash().GetHex());
+
+      wallet.BlockUntilSyncedToCurrentChain();
+    }
+    return block_hashes;
+  }
+
+  UniValue propose(const JSONRPCRequest &request) const override {
+    std::shared_ptr<CWallet> const wallet = GetWalletForJSONRPCRequest(request);
+    CWallet *const pwallet = wallet.get();
```

Comment:
NIT: I am not sure it is worth to extract this variable. Or maybe I am missing something?

Reply:
It can be simplified, yes.